src/wsproxy.ml

```diff
@@ -15,124 +15,110 @@
-         Lwt.return () in
-       let thread2 =
-         lwt_fd_enumerator fd (realunframe (writer (really_write localfd) "thread2")) >>= fun _ ->
-         Lwt.return () in
-       ignore_exn (fun () -> Lwt.join [thread1; thread2]) ()
-    )
-    (fun _ ->
-       ignore_exn (fun () -> Lwt_unix.close fd) () >>= fun () ->
-       ignore_exn (fun () -> Lwt_unix.close localfd) ())
-  >>= fun () ->
-  Printf.printf "FD %s pipe closed: %b %b\n" fdno Lwt_unix.(state fd == Closed) Lwt_unix.(state localfd == Closed)
-  |> Lwt.return
+      Lwt_log.warning_f "Unknown protocol, fallback to hybi10" >>= fun () ->
+      Lwt.return (wsframe, wsunframe) 
+  end >>= fun (frame,unframe) ->
+  with_open_connection_fd "localhost" localport ~callback:(fun localfd ->
+      let session_id = Uuidm.v `V4 |> Uuidm.to_string in
+      Lwt_log.debug_f "Starting proxy session %s" session_id >>= fun () ->
+      let thread1 =
+        lwt_fd_enumerator localfd (frame (writer (really_write fd) "thread1")) >>= fun _ ->
+        Lwt.return_unit in
+      let thread2 =
+        lwt_fd_enumerator fd (unframe (writer (really_write localfd) "thread2")) >>= fun _ ->
+        Lwt.return_unit in
+      (* closing the connection in one of the threads above in general leaves the other pending forever,
+       * by using choose here, we make sure that as soon as one of the threads completes, both are closed *)
+      Lwt.choose [thread1; thread2]
+      >>= fun () -> Lwt_log.debug_f "Closing proxy session %s" session_id)
+
 
 let handler sock msg =
-  Lwt_io.printf "Got msg: %s\n" msg >>= fun _ ->
+  Lwt_log.debug_f "Got msg: %s" msg >>= fun () ->
   match Re_str.split (Re_str.regexp "[:]") msg with
-  | [protocol;ty;sport] ->
-    let port = int_of_string sport in
-    proxy sock protocol ty port
+  | [protocol;_;sport]
   | [protocol;sport] ->
     let port = int_of_string sport in
-    proxy sock protocol "hybi10" port
-  | _ ->
-    let sid = Printf.sprintf "%d" (Obj.magic (Lwt_unix.unix_file_descr sock)) in
-    ignore_exn (fun () -> Lwt_unix.close sock) () >>= fun () ->
-    Printf.printf "Malformed message. Socket %s closed: %b\n" sid Lwt_unix.(state sock == Closed)
-    |> Lwt.return
+    proxy sock protocol port
+  | _ -> Lwt_log.warning "Malformed msg: not proxying"
+
 
-let _ =
-  if Array.length Sys.argv > 1
-  then Lwt.return (Websockets.runtest ())
-  else
-    begin
-      Lwt_daemon.daemonize ~stdout:`Dev_null ~stdin:`Close ~stderr:`Dev_null ();
-      let filename = "/var/run/wsproxy.pid" in
-      (try Unix.unlink filename with _ -> ());
-      Lwt_main.run begin
-        let pid = Unix.getpid () in
-        Lwt_io.with_file filename ~mode:Lwt_io.output (fun chan ->
-            Lwt_io.fprintf chan "%d" pid) >>= fun _ ->
-        start "wsproxy" handler
-      end
-    end
+let _ = 
+  (* Enable logging for all levels *)
+  Lwt_log.add_rule "*" Lwt_log.Debug;
+  Lwt_daemon.daemonize ~stdout:`Dev_null ~stdin:`Close ~stderr:`Dev_null ();
```

Comment:
But I think if we redirect sdout/err to /dev/null then we'll lose the errors reporting uncaught exceptions.

Reply:
I've found the same thing with xapi-nbd - I don't get log messages about uncaught exceptions.

Reply:
@mseri I actually decided to redirect stderr to syslog and stdout to null for xapi-nbd in the `.service` file, and it works quite well, I get logs like:
```
Sep 13 16:11:50 dt14 xapi-nbd[1133]: main: Starting xapi-nbd: port = '10809'; xen_api_uri = 'file:///var/xapi/xapi'; certfile = '/etc/xensource/xapi-ssl'; ciphersuites = '!EXPORT:RSA+AES128-SHA256' no_tls = 'false'
Sep 13 16:11:50 dt14 xapi-nbd[1133]: main: Initialising TLS
Sep 13 16:11:50 dt14 xapi-nbd[1133]: xapi-nbd: internal error, uncaught exception:
Sep 13 16:11:50 dt14 xapi-nbd[1133]: Ssl.Certificate_error
Sep 13 16:11:50 dt14 xapi-nbd[1133]: Raised by primitive operation at file "lwt/nbd_lwt_unix.ml", line 109, characters 2-43
Sep 13 16:11:50 dt14 xapi-nbd[1133]: Called from file "src/main.ml", line 142, characters 6-61
Sep 13 16:11:50 dt14 xapi-nbd[1133]: Called from file "src/main.ml", line 150, characters 19-73
Sep 13 16:11:50 dt14 xapi-nbd[1133]: Called from file "src/core/lwt.ml", line 672, characters 16-24
Sep 13 16:11:50 dt14 systemd[1]: xapi-nbd.service: main process exited, code=exited, status=1/FAILURE
Sep 13 16:11:50 dt14 systemd[1]: Unit xapi-nbd.service entered failed state.
Sep 13 16:11:50 dt14 systemd[1]: xapi-nbd.service failed.
```

Reply:
We do. I am still not 100% convinced of closing stderr for exactly that reason. Although good code should catch and report them instead of using stderr as a workaround

Reply:
@mseri So you reckon I should have a top-level wrapper for my whole program that catches, logs, and the reraises exceptions? (Right now I rely on OCaml printing uncaught exceptions to stderr.)

Reply:
I am not questioning their look, I had done it with stdout and stderr while debugging the old wsproxy. I think it's more polished to use  a global catcher for unexpected errors

Reply:
Makes sense. Then we will still log these unexpected errors even when we forget to redirect stderr to syslog. So its a kind of defence in depth, but for logging.

Reply:
@mseri I tried catching the exception at the top level, but now the backtrace became useless ;(
```
Sep 13 17:41:09 dt14 xapi-nbd[23859]: main: Starting xapi-nbd: port = '10809'; xen_api_uri = 'file:///var/xapi/xapi'; certfile = '/etc/xensource/xapi-ssl'; ciphersuites = '!EXPORT:RSA+AES128-SHA256' no_tls = 'false'
Sep 13 17:41:09 dt14 xapi-nbd[23859]: main: Initialising TLS
Broadcast message from systemd-journald@dt14 (Wed 2017-09-13 17:41:09 UTC):
xapi-nbd[23859]: main: Caught unexpected exception: Ssl.Certificate_error
Sep 13 17:41:09 dt14 xapi-nbd[23859]: main: Caught unexpected exception: Ssl.Certificate_error
Sep 13 17:41:09 dt14 xapi-nbd[23859]: main: Raised at file "map.ml", line 117, characters 16-25
Sep 13 17:41:09 dt14 xapi-nbd[23859]: main: Called from file "src/core/lwt.ml", line 161, characters 4-40
```
Do you know how to fix this?

Reply:
Oh, just read the Lwt manual, it says backtraces do not work with Lwt :(
http://ocsigen.org/lwt/manual/